### DISCUSSION TITLE
Setting different C/C++ Configuration based on directory path

### DISCUSSION THREAD

**Discussion by kmbgit** [microsoft/vscode-cpptools#8226] [Ideas]
The project I'm working on comprises modules build with a variety of C/C+  compiler toolchains, but the c_cpp_properties.json file only allow me to activate a single configuration at a time.  Is there a way to assign different configurations to different directories and have them work in parallel?  Currently when I swap from one toolchain to another I get masses of warnings for directory trees that are not built with the new toolchain.

I've come across the `configurationProvider` property but there' s no information on how to use it.

**Comment by sean-mcmanus** [comment]
Can you use a multi-root workspace, with one folder per directory that has a different configuration?

The configurationProvider property is meant to be used with a configuration provider extension, such as CMake Tools or Makefile Tools.

**Reply by kmbgit** [reply]
The project is set up with first-level folders for the different subsystems, so everything that uses one of the 4 different toolchains is in its own tree.  However, opening each of those 4 folders in separate VSCode instances is not ideal for a number of reasons - in particular it prevents me from easily searching across the whole project, it causes VSCode bloat because all the extensions have to be installed 4 times (I use Remote:SSH to run VSCode on my laptop against a dev server running Linux), and when I do a build from the top level it gives me problems navigating to the file causing an error or warning message.

**Reply by sean-mcmanus** [reply]
Having a workspace with 4 folders is different from having 4 VS Code instances, so you can search across all 4 folders: https://code.visualstudio.com/docs/editor/multi-root-workspaces. Have you tried that?

**Comment by ymeine** [comment]
Hi,

On my side I use a lot of workspaces for a lot of different projects, and each of them can have a lot of mixed content (many languages used, projects like PoCs, applications, tools, libraries, etc.). So eventually they are all multi-root (**lots** of roots), but I edit those roots very often based on what I want to focus on at a specific time.

My point is: having the specific configuration be based on a workspace folder is not viable for me, and I wish it could be per file system folder, independently of the workspace organization. That would also mean that the conf could provide a variable similar to `workspaceFolder`, and which would point to the folder where the configuration applies.

To illustrate my example: 

```
workspaceFolderA
    folderLib1
        lib1.h
    folderLib2
        .vscode\c_cpp_properties.json => includePaths points to ../folderLib1
        lib2.h
    folderApp
        .vscode\c_cpp_properties.json => includePaths points to ../folderLib2
workspaceFolderB
...
```

Unless I missed something, this is not achievable unless I do add `workspaceFolderA/folderLib2` and `workspaceFolderA/folderApp` as workspace roots, even if I wouldn't use them, knowing it would bloat my workspace structure.

Is there a hard limitation from VSCode regarding the config file being bound to a workspace folder?

**Reply by sean-mcmanus** [reply]
Configuration files in subfolders is just something we have not implemented. The recommended way to achieve something similar would be to use per-folder CMakeLists.txt with the CMake Tools extensions and configurationProvider set to that. Makefile Tools may be another alternative. compile_commands.json may also work if you have a way of generating it.

**Reply by ymeine** [reply]
Ok, then I understand that it is not a limitation, so that is something positive. I haven't taken the time yet to learn the contribution guidelines for this project, would you be able to roughly summarize me the procedure to suggest/vote features, if any? As for a potential PR, I'll see if I manage to free time (knowing the learning curve adds a lot).

And thanks for the tip involving CMake, I'll have a look at it. I tried to avoid having to learn/deal with CMake for now, as I went with other tool sets, but maybe I can just smartly use a subset to act solely as a configuration for the IDE!

Thanks again for your reply

**Reply by sean-mcmanus** [reply]
You can file a feature request issue, but I recall that one existed previously, and it may have been closed. Per-folder c_cpp_properties.json might not be possible to implement with only changes to our open source components.

**Reply by ymeine** [reply]
Alright, I will carefully search the issues to avoid any duplication, and I'll file a feature request if appropriate. And if you are right, I might even check the dependencies then.

Thanks for all the info!
